src/windows/hstring.rs

```diff
@@ -0,0 +1,102 @@
+use std::ptr::null_mut;
+
+use super::tz_on_calendar::TzOnCalendar;
+use super::winapi::windows_delete_string;
+use super::HRESULT;
+
+#[repr(C)]
+pub struct HstringHeader {
+    pub flags: u32,
+    pub len: u32,
+    pub padding1: u32,
+    pub padding2: u32,
+    pub ptr: *const u16,
+}
+
+#[repr(transparent)]
+pub struct Hstring(*mut HstringHeader);
+
+impl<'a> Hstring {
+    pub fn from_tz_on_calendar(instance: &'a TzOnCalendar) -> Result<Self, HRESULT> {
+        let mut string = null_mut();
+        // SAFETY: A `TzOnCalendar` is only ever created with a valid instance.
+        let result = unsafe {
+            let instance = instance.as_ptr();
+            ((**instance).GetTimeZone)(instance, &mut string)
+        };
+        if result < 0 || string.is_null() {
+            return Err(result);
+        }
+
+        Ok(Self(string))
+    }
+}
+
+impl Drop for Hstring {
+    fn drop(&mut self) {
+        // SAFETY: An `HString` is only ever created with a valid `HSTRING`.
+        let _ = unsafe { windows_delete_string(self) };
+    }
+}
+
+// The `w!(…)` and `h!(…)` macros were copied and adapted from
+// <https://github.com/microsoft/windows-rs/blob/cc2032e7462633e53e560ce1ee00a3ea9f87402e/crates/libs/windows/src/core/strings/literals.rs>.
+// © 2022 Microsoft, licensed under MIT OR Apache-2.0.
+
+/// Convert an ASCII string to a null-terminated `*const u16` wide string.
+/// The string must be ASCII and not contain NUL characters.
+macro_rules! w {
+    ($s:literal) => {{
+        const INPUT: &[u8] = $s.as_bytes();
+        const OUTPUT_LEN: usize = $s.len() + 1;
+        const OUTPUT: [u16; OUTPUT_LEN] = {
+            let mut buffer = [0; OUTPUT_LEN];
+            let mut pos = 0;
+            while pos < INPUT.len() {
+                // ensure that the input string is ASCII and does not contain any NUL characters
+                let _ascii_only = [0][if matches!(INPUT[pos], 1..=0x7f) { 0 } else { 1 }];
```

Comment:
The ASCII check isn't used for anything... but presumably it should be?

Reply:
Actually, no. I removed the UTF-8 → UCS2 conversion from the source, and the line is meant to throw an error at compile time if not all characters where in ASCII range. [Panicking in `const` contexts](https://caniuse.rs/features/const_panic) is only stable since 1.57, so I used the older workaround to access an array out of bounds, which is caught by the compiler. I only gave the result a name, so it would occur in the error message quoting the offending line